### ISSUE TITLE
Data representation/geometry of a GenomicAnnotation

### ISSUE THREAD

**Issue by nsheff** [fairtracks/fga-wg#3]
GenomicAnnotations have a genomic geometry; this is what Sveinung referred to as the "geometric type" of the data. Example: does the data represent Points in a genome (single nucleotides), or Regions in a genome (with start/end coordinates)? (Alternative name possibility: "data geometry"?)

One way I was thinking we could accommodate this is like so:

Each of these variables is boolean:
- are gaps allowed between elements?
- do elements have lengths?
	- if yes, are the lengths constant?
	- if yes, may elements overlap?
- do elements link to one another?
- do elements have values?

So, something like this:

```
geometry:
  gaps: true
  lengths: true
  lengths-constant: true
  lengths-overlap: true
  links: true
  values: true
```

**Comment by sveinugu** [comment]
The concept of genomic geomety in this issue is based on the "track type" concept from an old paper of mine from 2011, explained in this [slide show](https://drive.google.com/file/d/1gyXBJ35R7IKqh0Da59dNIG4BoHfQpTMZ/view?usp=drive_link) or this [section](https://fairtracks.net/tracks/#tracks-06-track-types) in the FAIRtracks web site. Illustration of the track types:

<img src="https://fairtracks.net/_nuxt/img/1525edf-800.png" width="400">

Also, the following table from that paper links the file formats, discussed in issue #1 , with the geometric types discussed in this issue:

![](https://github.com/user-attachments/assets/c8e86172-8167-4333-b72a-856cd1ed1818)

**Comment by sunbrn** [comment]
I agree with @nsheff on codifying the different types into a set of boolean variables, which makes it possible to have a simple two digit code for each type. On the other side, @sveinugu's paper (https://doi.org/10.1186/1471-2105-12-494) proposes a quite elegant representation.
I am not entirely sure the proposed schema has a 1:1 correspondence with the types from @sveinugu's paper. Can we sketch a mapping and see if everything is covered? When this is achieved, I would close the "Data geometry" issue

**Comment by rioualen** [comment]
Hi,
Following up on yesterday's meeting, I have started to revise the formats listed in the paper (Table 2) and the corresponding EDAM terms when available. I'm not finished yet but I've added it to our spreadsheet, if you want to take a look!

**Comment by sveinugu** [comment]
> So, something like this:
> 
> ```
> geometry:
>   gaps: true
>   lengths: true
>   lengths-constant: true
>   lengths-overlap: true
>   links: true
>   values: true
> ```

I think this is the way to go. The shorthand forms from the 2012 paper (or an updated version of this) can still be used, but they can be derived from your list of geometric properties

In addition, one can consider some other properties, e.g.

```
stranded: true
```

Which represents whether the region has strand information. I think this one is straightforward to introduce.

One could also potentially differentiate between different types of values, e.g. 

```
value_type: integer
value_dimension: 1
edge_weights: true
edge_weight_type: number
edge_weight_dimension: 1
```

These are from the specification of the GTrack format, also from my 2012 paper. However, even these properties are not comprehensive to cover all cases, as a file format may allow multiple types of values (e.g. alternative allele and quality score for the VCF file format). I the GTrack format, we assumed that one type of value was the dominant one, and for multi-dimensional values, we assume they are all of the same data type. 

To be fully comprehensive, I think one would really need to specify subschemas for the values and edge weight values, e.g.:

```
value_schema:
  - title: alt
    type: array
    items:
      type: string
  - tiitle: qual
    type: number
```

But this starts to become a file format parsing schema and seems WAY too detailed, especially given the complexity of some formats like VCF. So I don't think we should go in this direction. It also in many cases directly derivable from the file format, given we know the subtype (ref: the BED X+Y discussion in issue #1 ).

In conclusion, I think a straightforward boolean property to be added is whether the links also include edge weights:

```
links_have_weights: true
```

One could also include the GTrack properties more or less as they are, or collapse them to two, e.g.:

```
value_type: multiple
edge_weight_type: number
```

**Comment by sveinugu** [comment]
Two more possible properties from the GTrack spec:

```
circular_elements: false
```

True if any track element or bounding region cross the coordinate borders of a circular sequence

```
directed_edges: false
```

True if any edges are directed (was `undirected_edges` in GTrack for technical reasons)

**Comment by sunbrn** [comment]
All the properties you suggested, @sveinugu, seem relevant. It would help to have all of them in one place so we can reason on a complete proposal (and consider whether each of them would have a meaningful value for each known format).
Collecting the single suggestions above, I tried to make a full list. Please let me know if I understood all inputs correctly

geometry:
  - gaps: true
  - lengths: true
  - lengths-constant: true
  - lengths-overlap: true
  - edges: true
      - edges_have_weights: true
      - directed_edges: false
      - edge_weight_type: number
      - edge_weight_dimension: 1
  - values: true
      - value_title
      - value_type: integer/array
      - value_dimension: 1
      - value_items:
           - value_items_type
- stranded: true
- circular_elements: false

**Comment by sveinugu** [comment]
> values: true
> * value_title
> * value_type: integer/array
> * value_dimension: 1
> * value_items:
>   
>   * value_items_type

Here, you are going in the direction of subschemas for values, especially since you added `value_title`. My alternative is to simplify this to:

values: true
* value_type: integer/multiple

This relates to the measurement concept in the interpretation discussion in #2. Ift we for instance decide on a field denoting the interpretation of a measurement, `value_type` would denote the data type of this measurement.

I would think the question of supporting one "main" value relates to whether we should allow one or more "measurement" interpretations.

I will try to come up with a concrete, combined model that relates these thing.

**Comment by sveinugu** [comment]
In [Data Model v0.9](https://github.com/fairtracks/fga-wg/issues/13), geometric properties are covered by the `TrackGeometry` entity, stored in the field`GenomicAnnotationFile.track_geometry`.

Remaining subissues include:
- [ ] Consider whether the fields of TrackGeometry covers the needs discussed in this issue

**Comment by sveinugu** [comment]
I would prefer to keep the term **track geometry** rather than broaden this to **data geometry**. For me, the important point is not generic visualization, but that genomic annotation data has an intrinsic organization along a genome coordinate axis. That fundamentally one-dimensional structure is what gives rise to the geometric characteristics we care about here. In that sense, "track" is not just a genome browser term, but a useful way of referring to this particular class of coordinate-based genomic data.

If needed, we can clarify this in the definition/text, but I do not think broadening the term to "data geometry" would improve precision.
